pxd_fastpath.c

```diff
@@ -70,6 +69,94 @@ static void fastpath_flush_work(void) {
        }
 }
 
+int fastpath_adjust_fpthreads(int new_pxd_num_fpthreads)
+{
+	int old_pxd_num_fpthreads = pxd_num_fpthreads;
+	int node, cpu;
+	int rc = 0;
+
+	if (new_pxd_num_fpthreads > MAX_ALLOC_PXFP_WORKER_THREADS_PER_NODE) {
+		printk(KERN_WARNING"pxd_num_fpthreads(%d) over max limit(%d), reset to max\n", new_pxd_num_fpthreads, MAX_ALLOC_PXFP_WORKER_THREADS_PER_NODE);
+		new_pxd_num_fpthreads = MAX_ALLOC_PXFP_WORKER_THREADS_PER_NODE;
+	}
+	if (new_pxd_num_fpthreads == old_pxd_num_fpthreads) {
+		return 0;
+	}
+
+	for_each_online_node(node) {
+		struct pxfpcontext_per_node *c = &pxfpctxt[node];
+		const cpumask_t *cpumask = cpumask_of_node(node);
+		int active;
+
+		if (cpumask_empty(cpumask)) {
+			// NUMA node with no cpu's?! - skip it.
+			printk(KERN_NOTICE"skipping online numa node %d with no attached cpus\n", node);
+			continue;
+		}
+
+		active = 0;
+		// Count already created threads and create more if required
+		for_each_cpu(cpu, cpumask) {
+			struct kthread_worker* worker;
+			if (!cpu_online(cpu)) {
+				continue;
+			}
+
+			if (active == new_pxd_num_fpthreads) {
+				break;
+			}
+
+			// Thread is already created just increment count
+			// While decreasing active will reach new_pxd_num_fpthreads
+			// without creating new threads
+			if (c->fpworker[active] != NULL) {
+				active++;
+				continue;
+			}
+
+			// We need increase num threads, so create more threads
+			printk(KERN_NOTICE"Create thread on numa node %d cpu %d\n", node, cpu);
+			worker = kthread_create_worker_on_cpu(cpu, 0, "pxfpn%dc%d", node, cpu);
+			if (IS_ERR_OR_NULL(worker)) {
+				rc = PTR_ERR(worker);
+				goto out;
+			}
+			c->valid = true;
+			c->fpworker[active++] = worker;
+			if (fpdefault == NULL) {
+				fpdefault = worker;
+			}
+		}
+
+		// If active is less than new_pxd_num_fpthreads,
+		// num cpu is less than new_pxd_num_fpthreads and we cannot create more threads
+		if (active < new_pxd_num_fpthreads) {
+			new_pxd_num_fpthreads = active;
+		}
+	}
+
+	fastpath_map_workers();
+
+	printk(KERN_NOTICE"Updated fpthreads from %d to %d", pxd_num_fpthreads, new_pxd_num_fpthreads);
+	pxd_num_fpthreads = new_pxd_num_fpthreads;
+	return 0;
+
+out:
+	printk(KERN_ERR"fastpath fpthread init failure %d\n", rc);
+	for (node=0; node < MAX_NUMNODES; node++) {
+		struct pxfpcontext_per_node *c = &pxfpctxt[node];
+		if (c->valid) {
```

Comment:
how is now c->valid used? if the number of fp threads change. i do not see it modified anywhere.

Reply:
ok context is always valid... it is never invalid, it is the number of cpu threads under that may change.